drizzlepac/haputils/graph_utils.py

```diff
@@ -0,0 +1,258 @@
+
+            sourceCDS : ColumnDataSource object
+                The sourceCDS is an object based upon the Pandas dataframe holding the 
+                data of interest where the data can be referenced by column name
+
+            name : str, optional
+                Indentification of the glyph - useful when using HoverTool where the
+                value of the name is a dictionary key as the value itself is variable.
+
+            glyph_color : str, optional
+                The glyph_color of the "shape" glyphs can be specified via an input string 
+                (e.g. 'red', 'blue', 'colormap').  The term `colormap` specifies the
+                use of a pre-defined set of colors for the points derived from the 
+                `colormap` column in the input data `source`.  The default value for
+                this value is "colormap".
+    
+            legend_label : str, optional
+                Legend label to assign the particular glyph ONLY when the "glyph_color" 
+                argument is not set to "colormap".  If the "glyph_color" is "colormap", 
+                the "legend_label" argument is ignored.
+
+        """
+
+        # Set the required attributes
+        self.glyph_name = glyph_name
+        self.x = x
+        self.y = y
+        self.sourceCDS = sourceCDS
+    
+        # Check for optional attributes and use defaults as necessary.  The
+        # avaiable attributes are declared in the constructor.
+        self.size = data_dict.get('marker_size', self.size) 
+        self.fill_alpha = data_dict.get('fill_transparency', self.fill_alpha)
+        self.line_alpha = data_dict.get('line_transparency', self.line_alpha)
+
+        # If "glyph_color" is "colormap", then "legend_group" is the name of the column 
+        # from the "sourceCDS" to use for defining the legend for the colors used.
+        # The same colors should be assigned to all the same values of data from the
+        # column, for example, all 'ACS/WFC' data points from the "instrument" column.
+        # The "legend_label" is ignored when "glyph_color" is set to "colormap".
+        self.glyph_color = data_dict.get('glyph_color', self.glyph_color)
+        if self.glyph_color is 'colormap':
+            self.legend_group = 'gen_info.inst_det'
+            self.color = 'gen_info.colormap'
+        else:
+            self.color = self.glyph_color
+             
+        self.legend_label = data_dict.get('legend_label', self.legend_label)
+
+        # Dictionary of supported "shape" glyphs.  These are really references to
+        # the associated method names.
+        glyph_types = {'circle': HAPFigure.build_circle_glyph, 
+                       'square': HAPFigure.build_square_glyph,
+                       'triangle': HAPFigure.build_triangle_glyph}
+
+        glyph_types[glyph_name](self)
+
+
+    # "Shape" glyphs
+    def build_circle_glyph(self):
+
+        # hacky
+        if self.glyph_color is 'colormap':
```

Comment:
I am wondering whether this 'if...else' statement could simply define the values of 'legend_group' and 'legend_label' that would then get passed to a single call to 'self.fig.circle'.  Just a thought...

Reply:
You can only pass one of the possible "legend" parameters (of which there are three that exist) to the glyph creation functionality.  I am looking for alternatives to get rid of the hacky code.

Reply:
Not a blocker, merely a potential feature to be implemented or used.